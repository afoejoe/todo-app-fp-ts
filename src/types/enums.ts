export enum EFilter {
  ALL = "all",
  COMPLETED = "completed",
  INCOMPLETE = "incomplete",
}
