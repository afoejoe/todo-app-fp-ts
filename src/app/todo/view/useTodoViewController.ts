import { EFilter } from "@/types/enums";
import { useTodoViewModel } from "../view-model/useTodoViewModel";
import { useCallback, useEffect, useState } from "react";

export type UseTodoViewController = ReturnType<typeof useTodoViewController>;

export function useTodoViewController() {
  const [filter, setFilter] = useState(EFilter.ALL);

  const {
    getTodoList,
    footerText,
    data,
    handleDeleteTodo,
    handleToggleTodo,
    handleAddTodo,
    handleClearCompleted,
  } = useTodoViewModel({ filter });

  const handleFilterChange = useCallback((filter: EFilter) => {
    return () => {
      setFilter(filter);
    };
  }, []);

  useEffect(() => {
    getTodoList();
  }, [getTodoList]);

  return {
    filter,
    footerText,
    data,
    handleAddTodo,
    handleFilterChange,
    handleDeleteTodo,
    handleToggleTodo,
    handleClearCompleted,
  };
}
