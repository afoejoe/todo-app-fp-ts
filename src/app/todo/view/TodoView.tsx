import * as Todo from "@/components/todo";
import { useTodoViewController } from "./useTodoViewController";
import * as O from "fp-ts/Option";
import { pipe } from "fp-ts/lib/function";

export default function TodoView() {
  const {
    data,
    filter,
    footerText,
    handleAddTodo,
    handleDeleteTodo,
    handleToggleTodo,
    handleFilterChange,
    handleClearCompleted,
  } = useTodoViewController();

  return (
    <>
      <h1 className="text-3xl font-bold mb-4">Todo App</h1>
      {pipe(
        data,
        O.fold(
          () => <p>Loading...</p>,
          (data) => (
            <div className="flex flex-col w-full max-w-md px-4 py-8 bg-white rounded-lg shadow dark:bg-gray-800 sm:px-6 md:px-8 lg:px-10">
              <Todo.Header
                filter={filter}
                onAddTodo={handleAddTodo}
                onFilterChange={handleFilterChange}
              />

              <Todo.Body
                items={data}
                onDelete={handleDeleteTodo}
                onToggle={handleToggleTodo}
              />

              <Todo.Footer
                footerText={footerText}
                onClearCompleted={handleClearCompleted}
              />
            </div>
          ),
        ),
      )}
    </>
  );
}
