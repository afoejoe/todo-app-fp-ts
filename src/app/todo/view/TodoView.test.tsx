import { render } from "@testing-library/react";
import TodoView from "./TodoView";
import "@testing-library/jest-dom";
import { UseTodoViewController } from "./useTodoViewController";
import { EFilter } from "@/types/enums";
import { it } from "node:test";


let data: UseTodoViewController['data'] = []


jest.mock("./useTodoViewController", () => {
    return {
        useTodoViewController: () => ({
            data: data,
            footerText: "2 items left",
            handleAddTodo: jest.fn(),
            handleDeleteTodo: jest.fn(),
            handleToggleTodo: jest.fn(),
            handleFilterChange: jest.fn(),
            handleClearCompleted: jest.fn(),
            filter: EFilter.ALL,
        }),
    };
});



describe("TodoView Component", () => {
    it("shows loading... before data is fetched", () => {
        const { container, getByText, debug } = render(<TodoView />);
        expect(getByText("Loading...")).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it.skip("renders empty data correctly", () => {
        data = [
            {
                id: "1",
                title: "Todo 1",
                completed: false,
            },
            {
                id: "2",
                title: "Todo 2",
                completed: true,
            },
        ];


        const { getByText, queryByText, container, debug } = render(<TodoView />);
        expect(queryByText("Loading...")).not.toBeInTheDocument();

        expect(getByText("You haven't added any items")).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test.skip("renders some data correctly", () => {
        data = [
            {
                id: "1",
                title: "Todo 1",
                completed: false,
            },
            {
                id: "2",
                title: "Todo 2",
                completed: true,
            },
        ];

        const { container, debug } = render(<TodoView />);
        expect(container).toMatchSnapshot();
    })
});
