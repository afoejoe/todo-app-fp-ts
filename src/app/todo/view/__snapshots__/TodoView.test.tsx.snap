// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoView Component Rendering renders empty data correctly 1`] = `
<div>
  <h1
    class="text-3xl font-bold mb-4"
  >
    Todo App
  </h1>
  <div
    class="flex flex-col w-full max-w-md px-4 py-8 bg-white rounded-lg shadow dark:bg-gray-800 sm:px-6 md:px-8 lg:px-10"
  >
    <form
      class="flex mb-4"
    >
      <input
        class="flex h-10 w-full rounded-md border border-gray-200 bg-white px-3 py-2 text-sm ring-offset-white file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-gray-500 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-gray-800 dark:bg-gray-950 dark:ring-offset-gray-950 dark:placeholder:text-gray-400 dark:focus-visible:ring-gray-300 flex-1 mr-2"
        name="newTodoTitle"
        placeholder="New todo"
      />
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mb-4"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2 mr-2"
      >
        All
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 px-4 py-2 mr-2"
      >
        Completed
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 px-4 py-2"
      >
        Incomplete
      </button>
    </div>
    <ul
      class="space-y-2"
    />
    <div
      class="mt-4"
    >
      <p
        class="text-sm text-gray-500 dark:text-gray-400"
      >
        You haven't added any items
      </p>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2 mt-2"
      >
        Clear Completed
      </button>
    </div>
  </div>
</div>
`;

exports[`TodoView Component Rendering renders some data correctly 1`] = `
<div>
  <h1
    class="text-3xl font-bold mb-4"
  >
    Todo App
  </h1>
  <div
    class="flex flex-col w-full max-w-md px-4 py-8 bg-white rounded-lg shadow dark:bg-gray-800 sm:px-6 md:px-8 lg:px-10"
  >
    <form
      class="flex mb-4"
    >
      <input
        class="flex h-10 w-full rounded-md border border-gray-200 bg-white px-3 py-2 text-sm ring-offset-white file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-gray-500 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-gray-800 dark:bg-gray-950 dark:ring-offset-gray-950 dark:placeholder:text-gray-400 dark:focus-visible:ring-gray-300 flex-1 mr-2"
        name="newTodoTitle"
        placeholder="New todo"
      />
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mb-4"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2 mr-2"
      >
        All
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 px-4 py-2 mr-2"
      >
        Completed
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 px-4 py-2"
      >
        Incomplete
      </button>
    </div>
    <ul
      class="space-y-2"
    >
      <li
        class="dark:text-gray-900 flex items-center justify-between p-2 bg-gray-100 rounded"
      >
        <div
          class="flex items-center space-x-2"
          role="button"
          tabindex="0"
        >
          <button
            aria-checked="false"
            class="peer h-4 w-4 shrink-0 rounded-sm border border-gray-900 ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-gray-900 data-[state=checked]:text-gray-50 dark:border-gray-800 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 dark:data-[state=checked]:bg-gray-50 dark:data-[state=checked]:text-gray-900"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <span
            class=""
          >
            Todo 1
          </span>
        </div>
        <button
          aria-label="Delete"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 hover:bg-gray-100 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 w-10"
        >
          <svg
            class="lucide lucide-trash w-6 h-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 6h18"
            />
            <path
              d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
            />
            <path
              d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
            />
          </svg>
        </button>
      </li>
      <li
        class="dark:text-gray-900 flex items-center justify-between p-2 bg-gray-100 rounded"
      >
        <div
          class="flex items-center space-x-2"
          role="button"
          tabindex="0"
        >
          <button
            aria-checked="true"
            class="peer h-4 w-4 shrink-0 rounded-sm border border-gray-900 ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-gray-900 data-[state=checked]:text-gray-50 dark:border-gray-800 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 dark:data-[state=checked]:bg-gray-50 dark:data-[state=checked]:text-gray-900"
            data-state="checked"
            role="checkbox"
            type="button"
            value="on"
          >
            <span
              class="flex items-center justify-center text-current"
              data-state="checked"
              style="pointer-events: none;"
            >
              <svg
                class="lucide lucide-check h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20 6 9 17l-5-5"
                />
              </svg>
            </span>
          </button>
          <span
            class="line-through text-gray-500"
          >
            Todo 2
          </span>
        </div>
        <button
          aria-label="Delete"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 hover:bg-gray-100 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 w-10"
        >
          <svg
            class="lucide lucide-trash w-6 h-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 6h18"
            />
            <path
              d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
            />
            <path
              d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
            />
          </svg>
        </button>
      </li>
    </ul>
    <div
      class="mt-4"
    >
      <p
        class="text-sm text-gray-500 dark:text-gray-400"
      >
        You haven't added any items
      </p>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2 mt-2"
      >
        Clear Completed
      </button>
    </div>
  </div>
</div>
`;

exports[`TodoView Component Rendering shows loading... before data is fetched 1`] = `
<div>
  <h1
    class="text-3xl font-bold mb-4"
  >
    Todo App
  </h1>
  <p>
    Loading...
  </p>
</div>
`;
